README.md
# cobblestone

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2018 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-04-05
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-04-05
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/cobblestone/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/cobblestone/compare/0.1.0...0.1.1


test/cobblestone/core_test.clj
(ns cobblestone.core-test
  (:require [clojure.test :refer :all]
            [cobblestone.core :as pixel]
            [cobblestone.spec :as tiles]
            [cobblestone.json :refer [j2e]]
            [clojure.xml :as xml]
            [clojure.pprint :as pp]
            [clojure.edn :as edn]
            [clojure.string :as str]
            [clojure.spec.alpha :as s]
            [cheshire.core :as json]))

(defn- build-files [xml-file tile-doc]
  (let [svg (pixel/build-svg-from-tile-doc tile-doc)
        xml (->> svg
                 (xml/emit)
                 (with-out-str))]
    (spit (str "practice/" xml-file) xml)))

(def ^:private doc (edn/read-string (slurp "practice/practicetiles/pixel-tiles.edn")))

(def ^:private instruments (j2e (json/parse-string (slurp "practice/practicetiles/instruments.json") keyword)))

(deftest test-pixel-small
  (let [[tiles palettes size {:keys [single]}] doc]
    (build-files "one-tile.xml"
                 (into [tiles palettes] single))))

(deftest test-pixel-big
  (let [[tiles palettes size {:keys [simple-room]}] doc]
    (build-files "tiles.xml"
                 (into [tiles palettes size] simple-room))))

(deftest test-pixel-w-hall
  (let [[tiles palettes size {:keys [room-with-door]}] doc]
    (build-files "open.xml"
                 (into [tiles palettes size] room-with-door))))

(deftest test-explode-tile
  (let [full "aaaaaaaaaaaaaaaa|abbbbbbbbdcbbbbb|abcccdddddcbddbd|abccccddddcddddd|abdccccdddcddddd|abddcccccccccccc|abdddcccbbbbdcbb|abdddcccdddddcdd|abdddcbdccdddcdd|abbddcbdcccccccc|abdddcccccccbbbb|abdddcdddcccdddd|abbddcbddcbccccc|acccccbddcbdccdd|addddcbddcbdcccc|abdddcbddcbdcdcc"
        abbrev "a|ab7dcb|abc2d4cbddbd|abc3d3cd|abdc3d2cd|abddc|abd2c2b3dcb|abd2c2d4cd|abd2cbdccd2cd|abbddcbdc|abd2c6b|abd2cd2c2d|abbddcbddcbc|ac4bddcbdccd|ad3cbddcbdc|abd2cbddcbdcdc"
        smallest "a"
        all-As (str/join "|" (repeat 16 (apply str (repeat 16 "a"))))]

    (is (= full (pixel/explode-tile full)))
    (is (= full (pixel/explode-tile abbrev)))
    (is (= all-As (pixel/explode-tile all-As)))
    (is (= all-As (pixel/explode-tile smallest)))

    (pp/pprint (str/split (pixel/explode-tile "a||b||||a||||b||||a") #"[|]"))
    (pp/pprint (str/split (pixel/explode-tile "a|b||a||b||a||b||a||b||a") #"[|]"))))

(deftest test-pixel-stairs
  (let [[tiles palettes size {:keys [endless-stairs]}] doc
        single (into [tiles palettes size] endless-stairs)]
    (println (pr-str (s/explain-data ::tiles/tile-doc single)))
    (build-files "stairs.xml" single)))

(deftest test-pixel-instruments
  (let [[tiles palettes size {:keys [all-tiles]}] instruments
        single (into [tiles palettes size] all-tiles)]
    (println (pr-str (s/explain-data ::tiles/tile-doc single)))
    (build-files "instruments.xml" single)))


test/cobblestone/json_test.clj
(ns cobblestone.json-test
  (:require [clojure.test :refer :all]
            [clojure.edn :as edn]
            [cheshire.core :as json]
            [cobblestone.json :refer [j2e]]
            [clojure.spec.alpha :as s]
            [cobblestone.spec :as tiles]))

(def ^:private doc (edn/read-string (slurp "practice/practicetiles/pixel-tiles.edn")))

(deftest test-json
  (let [[tiles palettes _ lists] doc]
    (is (= doc (j2e (json/parse-string (json/generate-string [tiles palettes lists]) keyword))))
    (is (= nil (s/explain-data ::tiles/tile-doc-set doc)))
    ))


test/cobblestone/spec_test.clj
(ns cobblestone.spec-test
  (:require [clojure.test :refer :all]
            [clojure.spec.alpha :as s]
            [clojure.edn :as edn]
            [cobblestone.spec :as pixel]
            [clojure.string :as str]))

(def corner-tile
  (str/join
    "|"
    ["aaaaaaaaaaaaaaaa"
     "abbbbbbbbdcbbbbb"
     "abcccdddddcbddbd"
     "abccccddddcddddd"
     "abdccccdddcddddd"
     "abddcccccccccccc"
     "abdddcccbbbbdcbb"
     "abdddcccdddddcdd"
     "abdddcbdccdddcdd"
     "abbddcbdcccccccc"
     "abdddcccccccbbbb"
     "abdddcdddcccdddd"
     "abbddcbddcbccccc"
     "acccccbddcbdccdd"
     "addddcbddcbdcccc"
     "abdddcbddcbdcdcc"]))

(def edge-tile
  (str/join
    "|"
    ["aabddcbddcbdcdcc"
     "aabbdcdddcbdcdcc"
     "bbabdcccccbdcccc"
     "dbbabcbdddbdcdcc"
     "dddbacbddcbdcdcc"
     "cccccbbddccccdcc"
     "dddddbaddcddcdcc"
     "ddddddbadcbdcdbb"
     "dddddcddacbdcdcc"
     "cccccccccdbdcdcc"
     "bdcbbbbbbbadcdcc"
     "ddcddddddddbcdcc"
     "ccccccccccccdbcc"
     "dddddcdddddddbcc"
     "cccccccccccccccc"
     "cccccccccccccccc"]))

(def wall-tile
  (str/join
    "|"
    ["aaaaaaaaaaaaaaaa"
     "bdcbbbbbbdcbbbbb"
     "ddcbddbdddcbddbd"
     "ddcdddddddcddddd"
     "ddcdddddddcddddd"
     "cccccccccccccccc"
     "bbbbdcbbbbbbdcbb"
     "dddddcdddddddcdd"
     "dddddcdddddddcdd"
     "cccccccccccccccc"
     "bdcbbbbbbdcbbbbb"
     "ddcdddddddcddddd"
     "cccccccccccccccc"
     "dddddcdddddddcdd"
     "cccccccccccccccc"
     "cccccccccccccccc"]))

(def floor-tile
  (str/join
    "|"
    ["bcaabaabbcaabaaa"
     "abcaaacbbbcaaaaa"
     "aabcccbaaabcabaa"
     "aabbbbaaaaabcaba"
     "aabbaaaabacbbaaa"
     "acbaabaaacbbbcaa"
     "cbaabaaacbaaabcc"
     "bbabaaacbaaaaabb"
     "bbcaaacbaaaaaacb"
     "aabcccbaaaabacba"
     "aaabbbaaaabacbaa"
     "aaaabaababacbbaa"
     "abaabcaaaacbbbaa"
     "aabaabccccbaabca"
     "aaabaabbbbaaaabc"
     "caaaaaabbaaaaaab"]))

(deftest test-tile
  (is
    (= nil
       (s/explain-data
         ::pixel/tile
         corner-tile
         )))
  (is
    (= nil
       (s/explain-data
         ::pixel/tile-set
         {:corner corner-tile})))
  (is
    (= nil
       (s/explain-data
         ::pixel/tile-set
         {:corner corner-tile
          :wall wall-tile
          :floor floor-tile})))
  )

(deftest test-palette
  (is
    (= nil
       (s/explain-data
         ::pixel/palette
         [:lightgrey :grey :darkgrey])))

  (is
    (= nil
       (s/explain-data
         ::pixel/palette
         [:white :pink :purple :black])))

  (is
    (= nil
       (s/explain-data
         ::pixel/palette-set
         {:floor [:lightgrey :grey :darkgrey]
          :wall [:white :pink :purple :black]}
         )))

  )

(deftest test-tile-listing
  (is
    (= nil
       (s/explain-data
         ::pixel/tile-listing
         [:corner :wall :a/a]))))

(deftest test-tile-listing-1
  (is
    (= nil
       (s/explain-data
         ::pixel/tile-listing
         [:corner :wall :ab/a]))))

(deftest test-doc
  (let [tile-doc
        [{:corner corner-tile
          :wall wall-tile
          :floor floor-tile}
         {:floor [:lightgrey :grey :darkgrey]
          :wall [:white :pink :purple :black]}
         10
         [:corner :wall :a/a]
         [:corner :wall :flip-down :a/h]
         [:corner :wall :flip-over :j/a]
         [:corner :wall :flip-down :flip-over :j/h]
         [:wall :wall :bi/a]
         [:wall :wall :flip-down :bi/h]
         [:wall :wall :turn-left :a/bg]
         [:wall :wall :turn-left :flip-over :j/bg]
         [:floor :floor :bi/bg]]]
    (is
      (= nil
         (s/explain-data
           ::pixel/tile-doc
           tile-doc)))

    (println (pr-str (s/conform ::pixel/tile-doc tile-doc)))))

(def ^:private doc (edn/read-string (slurp "practice/practicetiles/pixel-tiles.edn")))

(deftest test-tiles
  (let [{:keys [cave-wall
                stone-edge
                brick-edge
                block
                floral-stone
                broken-floor
                cave-edge
                shiny-tile
                cave-floor
                stone-corner
                shuttered-door
                stone-floor
                stair-up
                cave-corner
                wet-floor
                stairs
                tile-floor
                wall-braziers
                standing-brazier
                stair-down
                brick-corner
                stone-tile
                statue
                bubble-tile
                stone-wall
                brick-wall]} (first doc)]

    (is (= nil (s/explain-data ::pixel/tile cave-wall)))

    (is (= nil (s/explain-data ::pixel/tile stone-edge)))

    (is (= nil (s/explain-data ::pixel/tile brick-edge)))

    (is (= nil (s/explain-data ::pixel/tile block)))

    (is (= nil (s/explain-data ::pixel/tile floral-stone)))

    (is (= nil (s/explain-data ::pixel/tile broken-floor)))

    (is (= nil (s/explain-data ::pixel/tile shiny-tile)))

    (is (= nil (s/explain-data ::pixel/tile cave-edge)))

    (is (= nil (s/explain-data ::pixel/tile stone-corner)))

    (is (= nil (s/explain-data ::pixel/tile shuttered-door)))

    (is (= nil (s/explain-data ::pixel/tile cave-floor)))

    (is (= nil (s/explain-data ::pixel/tile stone-floor)))

    (is (= nil (s/explain-data ::pixel/tile cave-corner)))

    (is (= nil (s/explain-data ::pixel/tile wet-floor)))

    (is (= nil (s/explain-data ::pixel/tile stairs)))

    (is (= nil (s/explain-data ::pixel/tile tile-floor)))

    (is (= nil (s/explain-data ::pixel/tile wall-braziers)))

    (is (= nil (s/explain-data ::pixel/tile standing-brazier)))

    (is (= nil (s/explain-data ::pixel/tile stair-down)))

    (is (= nil (s/explain-data ::pixel/tile stair-up)))

    (is (= nil (s/explain-data ::pixel/tile brick-corner)))

    (is (= nil (s/explain-data ::pixel/tile stone-tile)))

    (is (= nil (s/explain-data ::pixel/tile statue)))

    (is (= nil (s/explain-data ::pixel/tile bubble-tile)))

    (is (= nil (s/explain-data ::pixel/tile stone-wall)))

    (is (= nil (s/explain-data ::pixel/tile brick-wall)))

    ))

src/cobblestone/spec.clj
(ns cobblestone.spec
  (:require [clojure.spec.alpha :as s]
            [cobblestone.color :as colors]))

(def coordinate #"[a-z]([a-z])?")

(defn- coord-check [func]
  (fn [value]
    (let [str-val (func value)]
      (if-not (or (nil? str-val) (re-matches coordinate str-val))
        false
        (let [[low high] (map identity str-val)
              high (if (nil? high) low high)]
          (>= 0 (.compareTo (str low) (str high))))))))

(def tile-pattern-1 #"([a-z]([0-9])?)*[a-z]([|](([a-z]([0-9])?)*[a-z])?){0,15}")

(s/def ::color-component (s/and int? (partial <= 0) (partial >= 255)))

(s/def ::rgb-color
  (s/cat :r ::color-component
         :g ::color-component
         :b ::color-component))

(s/def ::hex-color (s/and keyword? #(re-matches #"[#][0-9A-F]{3}([0-9A-F]{3})?" (str %))))

(def color-names (set (mapv #(keyword (name %)) (keys colors/css))))

(def pantone-colors (set (mapv #(keyword (name %)) (keys colors/pantone))))

(s/def ::named-color color-names)

(s/def ::pantone-color pantone-colors)

(s/def ::color (s/alt :none #{:none}
                      :hex ::hex-color
                      :rgb ::rgb-color
                      :named ::named-color
                      :pantone ::pantone-color))

(s/def ::tile (s/and string? (partial re-matches tile-pattern-1)))

(s/def ::tile-name keyword?)

(s/def ::tile-set (s/map-of ::tile-name ::tile))

(s/def ::tile-loc (s/and keyword? (coord-check namespace) (coord-check name)))

(s/def ::palette (s/and vector?
                        (s/+ ::color)))

(s/def ::palette-name keyword?)

(s/def ::palette-set (s/map-of ::palette-name ::palette))

(s/def ::action #{:turn-right :turn-left :flip-down :flip-over})

(s/def ::tile-listing
  (s/and vector?
         (s/cat :name ::tile-name
                :palette-name ::palette-name
                :actions (s/* ::action)
                :locs (s/+ ::tile-loc))))

(s/def ::tile-doc
  (s/and vector?
         (s/cat :tiles ::tile-set
                :palettes ::palette-set
                :pixel-size (s/? int?)
                :list (s/+ ::tile-listing))))

(s/def ::doc-name keyword?)

(s/def ::tile-doc-set
  (s/and vector?
         (s/cat :tiles ::tile-set
                :palettes ::palette-set
                :pixel-size (s/? int?)
                :docs (s/map-of ::doc-name (s/and vector? (s/+ ::tile-listing))))))

src/cobblestone/server.clj
(ns cobblestone.server
  (:require [compojure.api.sweet :as sweet]
            [compojure.api.core :as api]
            [ring.util.http-response :as http]
            [ring.util.response :as resp]
            [compojure.route :as route]
            [org.httpkit.server :as server]
            [environ.core :refer [env]]
            [schema.core :as s]
            [cobblestone.core :as cob]
            [cheshire.core :as json]
            [cobblestone.json :refer [j2e]])
  (:gen-class))

(defn- build-response [code]
  (let [docs (j2e (json/parse-string code keyword))
        svgs (cob/build-svgs-from-tile-docs docs)]
    (vals svgs)))

(defn- build-app []
  (sweet/routes
    (sweet/api {}
               (api/context "/build" []
                            (sweet/resource {:description ""
                                             :post        {:summary    ""
                                                           :parameters {:body s/Any}
                                                           :responses  {200 {:schema s/Any}}
                                                           :handler    (fn [{body :body}]
                                                                         (let [code (slurp body)]
                                                                           (http/content-type
                                                                             (http/ok (build-response code))
                                                                             "application/json")))}}))
               (sweet/GET "/" [] (resp/redirect "/index.html")))
    (route/resources "/")
    (route/not-found "404 Not Found")))

(defn -main [& [port]]
  (let [my-app (build-app)
        port (Integer. ^int (or port (env :port) 5000))]
    (server/run-server my-app {:port port :join? false})))

resources/public/script/submit.js
function buildXML(node) {
    if ((typeof node) == "string") {
        return node;
    } else {
        node.attrs = node.attrs?node.attrs:{};
        node.content = node.content?node.content:[];
        var attrs = Object.entries(node.attrs).map(function(attr){
            return " " + attr[0] + "='" + attr[1] + "'";
        });
        var out = "<" + node.tag + attrs;
        if (node.content.length == 0) {
            out += "/>";
        } else {
            out += ">" + node.content.map(buildXML).join("") + "</" + node.tag + ">";
        }
        return out;
    }
}

function submitTileDoc(inputId,outputId) {
    var input = document.getElementById(inputId).value;
    var output = document.getElementById(outputId);
    var xhttp = new XMLHttpRequest();
    xhttp.onreadystatechange = function() {
        if (xhttp.readyState == 4 && xhttp.status == 200) {
            var svgData = JSON.parse(this.responseText);
            output.innerHTML = svgData.map(buildXML).join("");
        }
    };
    xhttp.open("POST", "/build", true);
    xhttp.send(input);
}

src/cobblestone/core.clj
(ns cobblestone.core
  (:require [clojure.string :as str]
            [clojure.spec.alpha :as s]
            [clojure.java.io :as io]
            [clojure.edn :as edn]
            [clojure.xml :as xml]
            [cobblestone.color :as colors]
            [cobblestone.spec :as pixel])
  (:import (clojure.lang ExceptionInfo)))

(defmulti ^:private process-color first)

(defmethod ^:private process-color :default [[_ str-val]] str-val)

(defmethod ^:private process-color :rgb [[_ {:keys [r g b]}]] (str "rgb(" r "," g "," b ")"))

(defmethod ^:private process-color :named [[_ name-key]] (name name-key))

(defmethod ^:private process-color :pantone [[_ pantone]] (name (get colors/pantone (keyword "pantone" (name pantone)))))

(defmethod ^:private process-color :none [_] nil)

(defmulti ^:private do-action
          (fn [_ action]
            action))

(defmethod ^:private do-action :default [point _] point)

(defmethod ^:private do-action :turn-right
  [{:keys [x y]} _] {:x (- 15 y) :y x})

(defmethod ^:private do-action :turn-left
  [{:keys [x y]} _] {:x y :y (- 15 x)})

(defmethod ^:private do-action :flip-down
  [{:keys [x y]} _] {:x x :y (- 15 y)})

(defmethod ^:private do-action :flip-over
  [{:keys [x y]} _] {:x (- 15 x) :y y})

(def ^:private printer-css
  "@media print {\n    svg {\n        page-break-after: always;\n        margin: 0.25in 0.5in 0.25in 0.5in;\n    }\n    @page {\n        size: landscape;\n        margin: 0;\n    }\n}")

(defn- index [c]
  (- (int c) 97))

(defn- optimize-tile [tile]
  (let [counts (reduce-kv #(assoc %1 %2 (count %3)) {} tile)
        bg (->> counts (sort-by second) last first)]
    {:bg bg
     :pixels (dissoc tile bg)}))

(defn- parse-tile [tile]
  (let [rows (mapv vector (range 16) (str/split tile #"[|]"))
        pixels (reduce
                 (fn [out [y ^String row]]
                   (reduce
                     (fn [out [x c]]
                       (conj out {:c c :x x :y y}))
                     out
                     (mapv #(vector %1 (index %2)) (range 16) row)))
                 []
                 rows)
        pixels (group-by :c pixels)]
    {:pixels (reduce-kv #(assoc %1 %2 (map (fn [p] (select-keys p [:x :y])) %3)) {} pixels)}))

(defn explode-tile [tile]
  (let [{:keys [out]} (reduce
                        (fn [{:keys [out]} row]
                          {:out
                           (if (empty? row)
                             (conj out (last out))
                             (let [row (reduce
                                         (fn [row-out c]
                                           (let [c (try (Integer/parseInt c) (catch Exception _ c))
                                                 c (if (int? c) (apply str (repeat c (last row-out))) c)]
                                             (str row-out c)))
                                         ""
                                         (map str row))
                                   diff (- 16 (count row))
                                   row (apply str row (repeat diff (last row)))]
                               (conj out row)))})
                        {:out []}
                        (str/split tile #"[|]"))
        diff (- 16 (count out))
        out (concat out (repeat diff (last out)))]
    (str/join "|" out)))

(defn- explode-svg [pixel-size defs uses width height]
  (let [tile-size (* 16 pixel-size)
        [width height] (map (partial * tile-size) [width height])]
    {:tag :svg
     :attrs {:width width :height height
             :xmlns "http://www.w3.org/2000/svg"
             :xmlns:xlink "http://www.w3.org/1999/xlink"}
     :content (into [{:tag :defs
                      :content (mapv (fn [{:keys [id bg pixels]}]
                                       (let [background (if (nil? bg) [] [{:tag :rect
                                                                           :attrs {:width tile-size
                                                                                   :height tile-size
                                                                                   :stroke "none"
                                                                                   :fill bg}}])]
                                         {:tag :g
                                          :attrs {:id id}
                                          :content (into background
                                                         (map (fn [{:keys [c x y]}]
                                                                {:tag :rect
                                                                 :attrs {:x (* x pixel-size)
                                                                         :y (* y pixel-size)
                                                                         :width pixel-size
                                                                         :height pixel-size
                                                                         :stroke "none"
                                                                         :fill c}})
                                                              pixels))}))
                                     defs)}]
                    (reduce
                      (fn [out {:keys [id x y]}]
                        (concat out
                                [{:tag :use
                                  :attrs {:x (* x tile-size)
                                          :y (* y tile-size)
                                          :xlink:href (str "#" id)}}
                                 {:tag :rect
                                  :attrs {:x (* x tile-size)
                                          :y (* y tile-size)
                                          :width tile-size
                                          :height tile-size
                                          :stroke "black"
                                          :stroke-width 1
                                          :fill "none"}}
                                 ]))
                      []
                      uses))}))

(defn- span-loc [loc]
  (let [[[min-x max-x] [min-y max-y]] (map (fn [func] (map index (func loc))) [namespace name])
        [max-x max-y] (mapv #(if (nil? %2) %1 %2) [min-x min-y] [max-x max-y])]
    (for [x (range min-x (inc max-x)) y (range min-y (inc max-y))] [x y])))

(defn- condenser [coll truncate-fn concat-fn]
  (loop [items coll
         out ""]
    (let [f (first items)
          r (rest items)
          n (count (take-while (partial = f) r))
          n (truncate-fn n)
          r (drop n r)
          f (if (or (zero? n) (empty? r)) (str f) (str f (concat-fn)))
          out (str out f)]
      (if (empty? r)
        out
        (recur r out)))))

(defn- condense-row [row]
  (condenser row #(if (< 9 %) 9 %) identity))

(defn- condense-rows [rows]
  (condenser rows identity #(repeat (inc %) "|")))

(defn condense-tile [tile]
  (let [tile (str/join "|" (str/split tile #"\n"))
        rows (str/split tile #"[|]")
        rows (mapv condense-row rows)]
    (condense-rows rows)))

(defn pivot-tile [tile]
  (let [tile (explode-tile (str/join "|" (str/split tile #"\n")))]
    (str/join "\n" (apply mapv #(apply str %&) (str/split tile #"[|]")))))

(defn- build-svg-from-exploded-tile-doc [tile-name {:keys [tiles palettes pixel-size list] :or {pixel-size 1}}]
  (let [tiles (reduce-kv #(assoc %1 %2 (parse-tile (explode-tile %3))) {} tiles)
        palettes (reduce-kv #(assoc %1 %2 (mapv process-color %3)) {} palettes)
        list (into (sorted-map) (mapv vector (range) list))
        {:keys [defs uses]} (reduce-kv
                              (fn [& n]
                                (let [[m k v] n
                                      {:keys [defs uses]} m
                                      index k
                                      {:keys [name palette-name actions locs]} v
                                      actions (if (empty? actions) [] actions)
                                      {:keys [pixels]} (get tiles name)
                                      palette (get palettes palette-name)
                                      {:keys [bg pixels]} (if (some nil? palette) {:pixels pixels} (optimize-tile pixels))
                                      pixels (reduce-kv #(assoc %1 %2 (mapv (fn [pixel] (reduce do-action pixel actions)) %3)) {} pixels)
                                      id (str tile-name index)
                                      bg (get palette bg)
                                      base (if (nil? bg) {} {:bg bg})]
                                  {:defs (conj defs (assoc base
                                                      :id id
                                                      :pixels (reduce-kv
                                                                #(let [color (get palette %2)]
                                                                   (if-not (nil? color)
                                                                     (concat %1 (mapv (fn [pixel] (assoc pixel :c color)) %3))
                                                                     %1))
                                                                [] pixels)))
                                   :uses (concat uses (reduce (fn [out loc]
                                                                (concat out (mapv (fn [[x y]] {:id id :x x :y y})
                                                                                  (span-loc loc))))
                                                              [] locs))}))
                              {:defs [] :uses []}
                              list)
        [width height] (mapv #(inc (apply max (mapv % uses))) [:x :y])]
    (explode-svg pixel-size defs uses width height)))

(defn build-svg-from-tile-doc [tile-doc]
  (when-let [error (s/explain-data ::pixel/tile-doc tile-doc)]
    (throw (ExceptionInfo. "Error in tile-doc" {:error error})))
  (build-svg-from-exploded-tile-doc "tile" (s/conform ::pixel/tile-doc tile-doc)))

(defn build-svgs-from-tile-docs [tile-docs]
  (when-let [error (s/explain-data ::pixel/tile-doc-set tile-docs)]
    (throw (ExceptionInfo. "Error in tile-doc" {:error error})))
  (let [tile-doc-set (s/conform ::pixel/tile-doc-set tile-docs)
        {:keys [tiles palettes pixel-size docs] :or {pixel-size 1} :as tiles-n-palettes} tile-doc-set
        tile-map (reduce-kv #(assoc %1 %2 {:tiles tiles :palettes palettes :pixel-size pixel-size :list %3}) {} docs)]
    (reduce-kv #(do (assoc %1 %2 (build-svg-from-exploded-tile-doc (name %2) %3))) {} tile-map)))

(defn build-html-from-tile-docs-text [in-str]
  (let [docs (edn/read-string in-str)
        svgs (build-svgs-from-tile-docs docs)
        build-page {:tag :html
                    :content [{:tag :head
                               :content [{:tag :style
                                          :content [printer-css]}]}
                              {:tag :body
                               :content (vals svgs)}]}]
    (with-out-str (xml/emit-element build-page))))

(defn -main [& [filename]]
  (let [path (.getParent (io/file filename))
        in-str (slurp path)
        build-html (build-html-from-tile-docs-text in-str)]
  (spit (.getAbsolutePath (io/file path "build.html")) build-html)))
